// @colyseus/schema/lib/types/HashMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ChangeTree_1 = require("../ChangeTree");
var HashMap = /** @class */ (function () {
    function HashMap(obj) {
        var _this = this;
        if (obj === void 0) { obj = {}; }
        this.length = 0;
        this.keys = [];
        this.values = [];
        this.$changes = new ChangeTree_1.ChangeTree();
        for (var key in obj) {
            this[key] = obj[key];
        }
        Object.defineProperties(this, {
            $changes: { value: undefined, enumerable: false, writable: true },
            onAdd: { value: undefined, enumerable: false, writable: true },
            onRemove: { value: undefined, enumerable: false, writable: true },
            onChange: { value: undefined, enumerable: false, writable: true },
            clone: {
                value: function (isDecoding) {
                    var cloned;
                    if (isDecoding) {
                        // client-side
                        cloned = Object.assign(new MapSchema(), _this);
                        cloned.onAdd = _this.onAdd;
                        cloned.onRemove = _this.onRemove;
                        cloned.onChange = _this.onChange;
                    }
                    else {
                        // server-side
                        var cloned_1 = new MapSchema();
                        for (var key in _this) {
                            if (typeof (_this[key]) === "object") {
                                cloned_1[key] = _this[key].clone();
                            }
                            else {
                                cloned_1[key] = _this[key];
                            }
                        }
                    }
                    return cloned;
                }
            },
            triggerAll: {
                value: function () {
                    if (!_this.onAdd) {
                        return;
                    }
                    for (var key in _this) {
                        _this.onAdd(_this[key], key);
                    }
                }
            },
            toJSON: {
                value: function () {
                    var map = {};
                    for (var key in _this) {
                        map[key] = (typeof (_this[key].toJSON) === "function")
                            ? _this[key].toJSON()
                            : _this[key];
                    }
                    return map;
                }
            },
            _indexes: { value: new Map(), enumerable: false, writable: true },
            _updateIndexes: {
                value: function () {
                    var index = 0;
                    var indexes = new Map();
                    for (var key in _this) {
                        indexes.set(key, index++);
                    }
                    _this._indexes = indexes;
                }
            },
        });
    }
    HashMap.prototype.add = function (key, value) {
        if (key === undefined && value === undefined) {
            return undefined;
        }
        var previous = undefined;
        //Are we replacing an existing element ?
        if (this.has(key)) {
            previous = this.values[this.index[key].data];
            this.values[this.index[key].data] = undefined;
            this.keys[this.index[key].key] = undefined;
        }
        else {
            this.length++;
        }
        //create a new index, this will replace existing one.
        this.index[key] = {
            data: this.values.push(value) - 1,
            key: this.keys.push(key) - 1
        };
        return previous;
    };
    HashMap.prototype.get = function (key) {
        if (this.has(key)) {
            return this.values[this.index[key].data];
        }
    };
    HashMap.prototype.has = function (key) {
        return (this.index[key] !== undefined);
    };
    HashMap.prototype.delete = function (key) {
        if (this.has(key)) {
            var previous = this.values[this.index[key].data];
            this.values[this.index[key].data] = undefined;
            this.keys[this.index[key].key] = undefined;
            this.length--;
            delete this.index[key];
            return previous;
        }
        else {
            return undefined;
        }
    };
    HashMap.prototype.forEach = function (fn, thisArg) {
        for (var i = 0; i < this.length; i++) {
            fn.call(thisArg, this.values[i], this.keys[i]);
        }
    };
    HashMap.prototype.clear = function () {
        this.values.length = 0;
        this.index = {};
        this.keys.length = 0;
        this.length = 0;
    };
    return HashMap;
}());
exports.HashMap = HashMap;
